### PR TITLE
Scheduling based on mix frames

### PR DESCRIPTION
**Pull request by reza-j**
This PR introduces scheduling of a quantum circuit based on mix frames in order to reduce the overall duration of the scheduled pulse sequence.

### CODE REVIEW

File: lib/Dialect/Pulse/Transforms/Scheduling.cpp [hunk 1]
```diff
@@ -113,14 +115,9 @@ void QuantumCircuitPulseSchedulingPass::scheduleAlap(
...
-      auto portsOrError =
-          PulseOpSchedulingInterface::getPorts(quantumGateSequenceOp);
-      if (auto err = portsOrError.takeError()) {
-        quantumGateSequenceOp.emitError() << toString(std::move(err));
-        signalPassFailure();
-      }
-      auto ports = portsOrError.get();
+      // find mix frame block ids of the quantum gate SequenceOp
+      std::unordered_set<uint> mixFramesBlockIds =
+          getMixFramesBlockIds(quantumGateCallSequenceOp);
```
Comment (taalexander) [inline lib/Dialect/Pulse/Transforms/Scheduling.cpp:120]: What is a mix frame block id?

File: lib/Dialect/Pulse/Transforms/Scheduling.cpp [hunk 2]
```diff
@@ -178,30 +175,38 @@ void QuantumCircuitPulseSchedulingPass::scheduleAlap(
...
-        portNameToNextAvailabilityMap.end()) {
-      if (portNameToNextAvailabilityMap[portName] < nextAvailableTimeOfAllPorts)
-        nextAvailableTimeOfAllPorts = portNameToNextAvailabilityMap[portName];
+int64_t QuantumCircuitPulseSchedulingPass::getNextAvailableTimeOfMixFrames(
+    std::unordered_set<uint> &mixFramesBlockIds) {
+  int64_t nextAvailableTimeOfAllMixFrames = 0;
+  for (auto id : mixFramesBlockIds) {
+    if (mixFrameToNextAvailabilityMap.find(id) !=
+        mixFrameToNextAvailabilityMap.end()) {
+      if (mixFrameToNextAvailabilityMap[id] < nextAvailableTimeOfAllMixFrames)
```
Comment (taalexander) [inline lib/Dialect/Pulse/Transforms/Scheduling.cpp:184]: It seems like this condition can be combined into a single boolean and `if`? (can use an intermediary value)

**Review by taalexander — CHANGES_REQUESTED** [review]
A few small questions

**Review by kitbarton — APPROVED** [review]
LGTM!

Please update the copyright dates in the files changed to include 2024 (for some reason GitHub won't allow me to add a comment to lines that are not changed).

**Review by taalexander — APPROVED** [review]
LGTM

**Pull request merged**